odoo/tools/image.py

```diff
@@ -32,89 +33,122 @@
+        - The original image ratio is preserved, unless `crop` is also given.
+        - If width or height is falsy, it will be computed from the other value
+            and from the ratio of the original image.
+        - If size is falsy or both width and height are falsy, no resize is done.
+    :type max_width: tuple (width, height)
+
+    :param verify_resolution: if True, make sure the original image size is not
+        excessive before starting to process it. The max allowed resolution is
+        defined by `IMAGE_MAX_RESOLUTION`.
+    :type verify_resolution: bool
+
+    :param quality: quality setting to apply.
+        - Ignored if image is not JPEG.
+        - 1 is worse, 95 is best. Default to 80.
+    :type quality: int
 
-    :param base64_source: the image base64 encoded
-    :type base64_source: string or bytes
+    :param crop: crop the image.
+        Instead of preserving the ratio of the original image, this will force
+        the output to take the ratio of the given `size`. Both `size` width
+        and height have to be defined for `crop` to work.
+        The value of `crop` defines where to crop: 'center', 'top', 'bottom'.
+        Default to 'center' if truthy.
+    :type crop: string
 
-    :param max_width: max width for the image. No resize if 0 given or if
-        the image is already smaller than the given value.
-    :type max_width: int
+    :param colorize: replace the trasparent background by a random color
+    :type colorize: bool
 
-    :param quality: quality setting to apply. Ignored if image is not JPEG.
-        1 is worse, 100 is best. Default to 80.
-        Set to 95 to keep the original quality.
-    :type quality: int
+    :param format: the output format. Can be PNG, JPEG or GIF.
+        Default to the format of the original image.
+        BMP is converted to PNG, other formats are converted to JPEG.
+    :type format: string
 
-    :return: optimized image base64 encoded
+    :return: image after the operations have been applied, base64 encoded
     :rtype: bytes
 
-    :raise: ValueError if image too large
+    :raise: ValueError if `verify_resolution` is True and the image is too large
+    :raise: binascii.Error: if the base64 is incorrect
+    :raise: OSError if the image can't be identified by PIL
     """
+    if not base64_source:
+        return False
     if base64_source[:1] == b'P':
         # don't process SVG
         return base64_source
 
     image = base64_to_image(base64_source)
 
     w, h = image.size
-    if w * h > IMAGE_MAX_RESOLUTION:
+    if verify_resolution and w * h > IMAGE_MAX_RESOLUTION:
         raise ValueError(_("Image size excessive, uploaded images must be smaller than %s million pixels.") % str(IMAGE_MAX_RESOLUTION / 10e6))
 
     # get the format of the original image (must be done before resize)
-    opt = {'format': image.format}
+    format = (format or image.format).upper()
+    if format == 'BMP':
```

Comment:
rename format to avoid erase: <built-in function format>

Reply:
output_format ? output

Reply:
Ok for `output_format`